pyxem/utils/segment_utils.py

```diff
@@ -0,0 +1,313 @@
+            index = np.transpose(maxi_coord1)[i]
+            label_value = labels_check[index[0], index[1]]
+            if len(labels_check[labels_check == label_value]) < min_size:
+                delete_indices.append(i)
+                local_maxi[index[0], index[1]] = False
+        maxi_coord1 = np.delete(maxi_coord1, delete_indices, axis=1)
+
+    # Cluster the maxima by DBSCAN based on min_distance. For each
+    # cluster, only the maximum closest to the average maxima position is
+    # used as a marker.
+    if min_distance > 1 and np.shape(maxi_coord1)[1] > 1:
+        clusters = DBSCAN(eps=min_distance, metric='euclidean', min_samples=1,
+                          ).fit(np.transpose(maxi_coord1))
+        local_maxi = np.zeros_like(local_maxi)
+        for n in np.arange(clusters.labels_.max() + 1):
+            maxi_coord1_n = np.transpose(maxi_coord1)[clusters.labels_ == n]
+            com = np.average(maxi_coord1_n, axis=0).astype('int')
+            index = distance_matrix([com], maxi_coord1_n).argmin()
+            index = maxi_coord1_n[index]
+            local_maxi[index[0], index[1]] = True
+
+    # Use the resulting maxima as markers. Each marker should have a
+    # unique label value. For each maximum, generate markers with the same
+    # label value in a radius given by marker_radius centered at the
+    # maximum position. This is done to make the segmentation more robust
+    # to local changes in pixel values around the marker.
+    markers = label(local_maxi)[0]
+    if marker_radius >= 1:
+        disk_mask = disk(marker_radius)
+        for mm in np.arange(1, np.max(markers) + 1):
+            im = np.zeros_like(markers)
+            im[np.where(markers == mm)] = markers[np.where(markers == mm)]
+            markers_temp = convolve2d(im, disk_mask, boundary='fill',
+                                      mode='same', fillvalue=0)
+            markers[np.where(markers_temp)] = mm
+    markers = markers * mask
+
+    # Find the edges of the VDF image using the Sobel transform.
+    elevation = sobel(vdf_temp)
+
+    # 'Flood' the elevation (i.e. edge) image from basins at the marker
+    # positions. Find the locations where different basins meet, i.e.
+    # the watershed lines (segment boundaries). Only search for segments
+    # (labels) in the area defined by mask.
+    labels = watershed(elevation, markers=markers, mask=mask)
+
+    sep = np.zeros((np.shape(vdf_temp)[0], np.shape(vdf_temp)[1],
+                    (np.max(labels))), dtype='int32')
+    n, i = 1, 0
+    while (np.max(labels)) > n - 1:
+        sep_temp = labels * (labels == n) / n
+        sep_temp = np.nan_to_num(sep_temp)
+        # Discard a segment if it is too small, or else add it to the list
+        # of separated segments.
+        if ((np.sum(sep_temp, axis=(0, 1)) < min_size)
+                or (max_size is not None
+                    and np.sum(sep_temp, axis=(0, 1)) > max_size)):
```

Comment:
This condition needs a test.

Reply:
@tinabe The multiple conditional statements here make this bit of code quite difficult to read and understand fully - can you simplify it? If not can you write a test that uses this part of the code?